### DISCUSSION TITLE
Am I using nuqs with nextjs wrong?

### DISCUSSION THREAD

**Discussion by J4v4Scr1pt** [47ng/nuqs#778] [Q&A] [answered]
Nuqs version: 2.2.2

Don't get me wrong, I <ins>really really</ins> like nuqs and see the use of it! It's super awesome!

But I just implemented it a big project. Replacing the previous code using nextjs suggestion on [using native history API](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api) to avoid unnecessary  re-renders of RSC. Nuqs turned the code to be much cleaner and easier to work with.

However now we started to get a lot of server re-render which is not what I was expecting because of [this ](https://nuqs.47ng.com/docs/options#shallow)in the docs.

I'm now considering removing it to go back to the previous implementation we had. 

But I don't want to do that, so before I do it what am I missing here? 
Is there some setting that I should turn off/on? 

How is it working under the hood? Does it use the NextJs suggested way of working with the native history API as stated in docs with shallow?

Enlighten a stupid person as me please 😅 

As an example, I have a simple dropdown with items that you can select, underneath is a list things based on the items Id. 
When I select an item in the list I change the URL to reflect the new Id and trigger Tanstack to fetch the new data based on the items Id. 
That's when I noticed the re-renders. I then removed Tanstack and all other code except the dropdown, but the re-renders still happens when I select the Id.

Code with the list to show, cleaned to show only the Id atm:
```
'use client';

import { useQueryState } from 'nuqs';

export function ListOfItems() {
	const [selectedId] = useQueryState('selectedId');
// tanstack trigger will happen here
return <div>{selectedId}</div>
}
```
The select:
```
'use client';

import { useQueryState } from 'nuqs';

export function Header({defaultSelectedId}:{defaultSelectedId:string}) {
	const [_, setSelectedId] = useQueryState('selectedId', {
		defaultValue: defaultSelectedId,
	});

	return (
	
			<Select
				size="sm"
				variant="bordered"
				radius="sm"
				color="default"
				onChange={(i) => setSelectedId(i.target.value)}
				disallowEmptySelection>
				{(item) => (
					<SelectItem key={item.Id} value={item.Name}>
						{item.Name}
					</SelectItem>
				)}
			</Select>
	
	);
}
```

**Comment by franky47** [comment]
> However now we started to get a lot of server re-render which is not what I was expecting because of [this](https://nuqs.47ng.com/docs/options#shallow)in the docs.

Are you sure your server-rerenders are coming from nuqs? Because as you said, and as it is, the default internal behaviour is to use the History API as recommended by Next.js for client-only updates of the URL.

This **will** cause re-renders on the client (at least two as the URL udpate is asynchronous, so one for the optimistic state update and one when the URL finishes updating to get back in sync), but it should **not** trigger any re-renders on the server on its own, unless you opt-in to do so with `shallow: false`.

**Reply by J4v4Scr1pt** [reply]
Thx for clarifying this, then I have understood everything correctly.
Then I do not understand why the re-render happens. 
Right now I put a console.log in all RSCs that shows on the page, side menu, top navigation  etc.
The console.log only popup when I use nuqs setQueryState but not native window.history.replaceState. That's why I thought this was the issue.
I can check a bit more to try and understand where the re-renders come from.

**Reply by franky47** [reply]
If you have a minimal reproduction example, I can take a look as well.

**Reply by J4v4Scr1pt** [reply]
Thx m8!
I can see if I can create one later if I don't find anything. 
Our project is big and to have roughly the same structure could take some time to put together in an minimal example 😸

**Reply by J4v4Scr1pt** [reply]
First off, I apologize for the bad title I now realize. Will change it.

It's a bit inconsistent for me, it triggers re-renders sometimes and sometimes not.

To take the the example from my initial post.
If I select an item in the list that has not been selected before it does **not** trigger RSC re-render. Then if I select an already selected item the re-render accurse, as if the value was cached somewhere and told NextJs to re-fetch the RSC's.

**Reply by J4v4Scr1pt** [reply]
I cant wrap my head around this. If use regular 
`window.history.replaceState(null, '', '?selectedId=${id}')} `to set the query param and use `useQueryState `to get the param everything works great. No re-rerenders or anything. But as soon as I use the setter from useQueryState I get the re-renders again.

I don't see any performance issues having the useQueryState, but very annoying to have the extra roundtrips to the server because of client side URL changes. It kind of destroys the use of having the URL as a state manager.

**Reply by franky47** [reply]
What version of Next.js are you on? That sounds a lot like a client cache thing (the 30s for which a page is considered fresh on the client for a given URL, before it gets re-fetched). But that shouldn't do anything different whether you use nuqs or the history API.

**Reply by J4v4Scr1pt** [reply]
I'm using latest Nextjs 14 can't upgrade to 15 yet because of packages still not supporting react 19.
Very interesting, have I missed this information.. I know there is a lot to NextJs cache.. 😅.

**Reply by J4v4Scr1pt** [reply]
Tried to debug together with v0 and co-pilot. Both of them suggested that the wrapping provider could be the issue. I know to take their answers with a grain of salt.. or full bucket.. 😅 .
How does NuqsAdapter for the app router work?

**Reply by franky47** [reply]
The adapter is a Context provider that passes a hook function (implementing the router interface with each framework) to the useQueryState(s) hooks. It doesn't do anything else by itself.

Are you on the app or pages router? Edit: app router, so technically you're already using React 19 even in Next.js 14, they embed their own version of React canary in the app router (the version you indicate in your package.json doesn't matter).

**Reply by J4v4Scr1pt** [reply]
Thanks for all the fast response! 
That's true what you say about react19. and yes only app router. But if it's just a regular Context provider then AI is lying.. 😄.
I will keep testing and get back again.

**Reply by franky47** [reply]
Yeah you might want to not trust AI on a piece of code that is less than a month old 🙃

https://github.com/47ng/nuqs/blob/next/packages/nuqs/src/adapters/next/impl.app.ts

**Reply by J4v4Scr1pt** [reply]
hehe I try to not trust AI as much as possible.. 😅.

But I do see that it's using the `router `to do the changes. Could this affect the re-render?
Have you considered using only the native `window `object as they suggest in the docs?
Not to replace `router `but rather have an setting on the options, something like "useNative" or similar? 

I saw tough when I tested with `window.history` the `useQueryState ` first got the previous value as a change before it got the correct new value.

Sorry if sound stupid or something I just brainstorming here. Nuqs is superb lib, like it a lot! 💪

**Reply by J4v4Scr1pt** [reply, marked as answer]
We can close this.

I found the issue it has nothing to do with neither Tanstack or nuqs... just my stupid *ss.. 😭.

In the docs for NextJs it does not say anything about you cant use ServerActions to fetch data. Even some youtubers have suggested doing this.

But I know better now, after I converted all my GET serverActions to be route-handlers all is good.

Thanks again for all your awesome responses!

**Reply by franky47** [reply]
Good to know, though this might change soon, as server actions were renamed to server functions in the React docs, so I guess the React core team has plans for them to be more than just POST mutations in the future, and finally erase the API routes paradigm.

**Reply by J4v4Scr1pt** [reply]
Let's hope so, because "serverActions" are so neat to write 😄
